File: java/740_Delete_and_Earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        if(nums == null) return 0;
        int max = Integer.MIN_VALUE;

        for(int num : nums){
            max = (num > max) ? num : max;
        }

        int[] sumNums = new int[ max + 1];
        
        for(int num : nums){
            sumNums[num] += num;
        }
        
        int pre = 0 , cur = 0;
        for(int sum : sumNums){
            int temp = pre;
            pre = cur;
            cur = Math.max( temp+sum, cur );
            // pre = cur;
        }
        
        return cur;

    }
}

File: python/740_Delete_and_Earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        val = [0 for _ in range(max(nums)+1)]
        for n in nums:
            val[n] += n
        
        def rob(nums):
            pre, cur = 0,0
            for num in nums:
                pre,cur = cur,max(cur,pre+num)
            return cur
        
        return rob(val)
